Add WPA-Enterprise support

DurandA opened the pull request:
Early support of WPA-Enterprise. This PR is not yet complete and should not be merged.

DurandA commented:
Hi @shmuelzon!

Thank you for your project. I tried it again since last year and stability improved drastically.

I am working on adding WPA-Enterprise support and have a few questions/remarks:
* Would you be interested in merging this feature when it will be complete?
* Do you have any remarks regarding to what was already commited?
* What do you think about the `eap` section structure in the configuration file?

shmuelzon commented:
Hey Arnaud,

Thank you for the kind words.
Generally speaking, assuming it works, I see no reason why it shouldn’t be merged.
As for the code itself, I’ll take a look when I have some free time.
Regarding the configuration, those are a lot of fields but I assume they’re all needed. I’ll look a bit into it to understand what each parameter is.

Goo fay!

DurandA commented:
Yay!

It currently lacks support for certificates and should be tested properly. I also want to ensure that EAP functions are defined in their own section instead of being scattered in _config.c_.

I was able to connect to my university network using PEAP. Unfortunately, I don't have access to enterprise network with TLS or TTLS authentication so I won't be able to test it myself.

DurandA commented:
I am now happier with the PR.

Here is the configuration I used for testing:
```json
"wifi": {
  "ssid": "secure-unifr",
  "eap": {
    "method": "PEAP",
    "ca_cert": "/certs/QuoVadis_Root_CA_2.pem",
    "username": "MY_USERNAME",
    "password": "MY_PASSWORD"
  }
}
```

CA certificates are optional. Client certificates are not used with my network and I cannot test it. Since this implementation is very close to the [wpa2_enterprise example](https://github.com/espressif/esp-idf/tree/release/v3.2/examples/wifi/wpa2_enterprise) I don't expect any bad surprise.

In the README, `data/certs/my_cert.pem` is given as a path example for certificate. However, it seems that only files in the root (`data`) folder are added to SPIFFS. Should I update the Makefile?

shmuelzon commented on README.md:
```diff
@@ -93,12 +93,28 @@ The `wifi` section below includes the following entries:
+  * `identity` - The EAP identity
+  * `ca_cert`, `client_cert`, `client_key` - Full path names, including a
+    leading slash (/), of the certificate/key file (in respectively PEM, CRT and
+    KEY format) stored under the data folder.
```
Correct me if I'm wrong, but aren't they all PEM format?

DurandA replied:
You are right, crt (e.g. [wpa2_client.crt](https://github.com/espressif/esp-idf/blob/release/v3.2/examples/wifi/wpa2_enterprise/main/wpa2_client.crt)) and key [wpa2_client.key](https://github.com/espressif/esp-idf/blob/release/v3.2/examples/wifi/wpa2_enterprise/main/wpa2_client.key) files are in PEM format, just with [different extension](https://serverfault.com/a/9717/405370). I'll fix the README.

shmuelzon commented on main/wifi.c:
```diff
@@ -77,14 +78,60 @@ static esp_err_t event_handler(void *ctx, system_event_t *event)
 
     ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
     ESP_ERROR_CHECK(esp_wifi_set_config(ESP_IF_WIFI_STA, &wifi_config));
+    if (eap_method) {
```
Please move the bracket to a new line to fit the rest of the code

shmuelzon commented on main/wifi.c:
```diff
@@ -77,14 +78,60 @@ static esp_err_t event_handler(void *ctx, system_event_t *event)
+    if (eap_method) {
+        ESP_ERROR_CHECK(esp_wifi_sta_wpa2_ent_set_ca_cert((uint8_t *)ca_cert,
+            ca_cert ? strlen(ca_cert) : 0));
+        ESP_ERROR_CHECK(esp_wifi_sta_wpa2_ent_set_cert_key((uint8_t *)client_cert,
```
Why not check for `client_cert && client_key` before calling `esp_wifi_sta_wpa2_ent_set_cert_key()`?
I think it will be a bit more readable. Same for line 119 above.

DurandA replied:
I will check for `client_cert` before calling `esp_wifi_sta_wpa2_ent_set_cert_key()`. However, `client_key` is optional as it is used when the key from the client certificate is encrypted.

Should I check for `client_key` and write two different calls to `esp_wifi_sta_wpa2_ent_set_cert_key()` considering if `client_key` is present or not?

Another option is to drop the feature entirely since it is possible to put the decrypted key in the certificate directly.

shmuelzon replied:
> `client_key` is optional as it is used when the key from the client certificate is encrypted.

Are you sure about that? This should be the private key corresponding with the public key set in the certificate, not a passphrase used to protect the private key. I don't think this is optional at all.

DurandA replied:
> Are you sure about that? This should be the private key corresponding with the public key set in the certificate, not a passphrase used to protect the private key.

You are indeed right as the `private_key_password` is another field: https://github.com/espressif/esp-idf/blob/7dd49231903643d641ed4ccaff01422a9db51fb2/components/esp32/include/esp_wpa2.h#L167-L172.

> I don't think this is optional at all.

I would expect it to be optional if the private key is included in the client certificate. E.g., [wpa2_client.pem](https://github.com/espressif/esp-idf/blob/release/v3.2/examples/wifi/wpa2_enterprise/main/wpa2_client.pem) from the example contains the private key (which is encrypted—denoted with `BEGIN ENCRYPTED PRIVATE KEY` or `Proc-Type: 4,ENCRYPTED` metadata). Whenever the ESP32 extract the private key from the client certificate will remain a mystery until tested or asking Espressif since the underlying implementation is a binary blob (same is true for [SSL](https://github.com/espressif/esp-idf/blob/7dd49231903643d641ed4ccaff01422a9db51fb2/components/tcp_transport/transport_ssl.c#L172-L188)).

DurandA replied:
I asked the question on [ESP32 Forum](https://www.esp32.com/viewtopic.php?f=13&t=12233) since that it might be useful to other people and testing it is cumbersome (especially for EAP).

shmuelzon replied:
Thanks for checking that out. In their example, they still pass a key file for the connection so it does seem odd.
If it's OK by you, I would rather wait for an answer on the forum before merging.

shmuelzon commented on main/wifi.c:
```diff
@@ -77,14 +78,60 @@ static esp_err_t event_handler(void *ctx, system_event_t *event)
+                strlen(eap_identity)));
+        if (eap_method == EAP_PEAP || eap_method == EAP_TTLS) {
+            ESP_ERROR_CHECK(esp_wifi_sta_wpa2_ent_set_username((uint8_t *)eap_username,
+                strlen(eap_username)));
```
What if `eap_username` or `eap_password` weren't in the configuration file? Won't this crash?

DurandA replied:
This will indeed crash. However and if I'm not mistaken, both `eap_username` and `eap_password` are required when the method is Tunnelled TLS or Protected EAP.

shmuelzon replied:
You know that, and I know that (now :)), but who know what the user will do. I rather they see an error message that it can't connect to WiFi than crash without a warning.

shmuelzon commented:
> In the README, data/certs/my_cert.pem is given as a path example for certificate. However, it seems that only files in the root (data) folder are added to SPIFFS. Should I update the Makefile?

Are you sure about that?
```bash
$ mkdir data/certs
$ touch certs/my_cert.pem
$ make $(PWD)/build/spiffs.bin
Toolchain path: /Users/assaf/Downloads/esp/xtensa-esp32-elf/bin/xtensa-esp32-elf-gcc
Toolchain version: crosstool-ng-1.22.0-80-g6c4433a
Compiler version: 5.2.0
/config.json
/certs/my_cert.pem
```
AFAIK, SPIFFS doesn't support folders. Instead, to simulate it, a file called "/certs/my_cert.pem" is created. But still shouldn't affect anything and I've tested the MQTT certificates inside a `certs` directory and they loaded as expected.

DurandA commented:
Thanks for the review!

I added some of the proposed changes to 0ff0ed6. Please see my comments above for the elements I didn't change.

> AFAIK, SPIFFS doesn't support folders. Instead, to simulate it, a file called "/certs/my_cert.pem" is created. But it still shouldn't affect anything and I've tested the MQTT certificates inside a `certs` directory and they loaded as expected.

I will test it again. I added some debug statements around there to check if the certificate is loaded:
https://github.com/shmuelzon/esp32-ble2mqtt/blob/e5417e2d2f0cc5cf680432f94e654820122688fc/main/config.c#L261-L262
I may have screwed something, I will report back if it still fails.

shmuelzon commented on main/wifi.c:
```diff
@@ -77,14 +78,63 @@ static esp_err_t event_handler(void *ctx, system_event_t *event)
     ESP_ERROR_CHECK(esp_wifi_set_config(ESP_IF_WIFI_STA, &wifi_config));
+    if (eap_method)
+    {
+        ESP_ERROR_CHECK(esp_wifi_sta_wpa2_ent_set_ca_cert((uint8_t *)ca_cert,
```
Please check for `ca_cert` before calling `esp_wifi_sta_wpa2_ent_set_ca_cert()`.

DurandA replied:
Sorry about that!

shmuelzon commented on main/wifi.c:
```diff
@@ -77,14 +78,63 @@ static esp_err_t event_handler(void *ctx, system_event_t *event)
+    {
+        ESP_ERROR_CHECK(esp_wifi_sta_wpa2_ent_set_ca_cert((uint8_t *)ca_cert,
+            ca_cert ? strlen(ca_cert) : 0));
+        if (client_cert)
```
Since this is a multi-line if now, it should be put inside curly-braces.

shmuelzon commented on main/wifi.c:
```diff
@@ -77,14 +78,63 @@ static esp_err_t event_handler(void *ctx, system_event_t *event)
+            ESP_ERROR_CHECK(esp_wifi_sta_wpa2_ent_set_cert_key((uint8_t *)client_cert,
+                strlen(client_cert), (uint8_t *)client_key,
+                client_key ? strlen(client_key) : 0, NULL, 0));
+        if (eap_identity)
```
Since this is a multi-line `if`, it should be put inside curly-braces.

shmuelzon commented:
Sorry to being so pedantic, but there's just one more thing I noticed.
My (non-documented) coding convention is to limit lines to 80 characters, could you please fix that as well?
After we hear back from the forum, I'll merge this PR.

shmuelzon commented:
Looks like we're not going to get a response from Espressif.
I'll merge this for now and if we ever discover that the key isn't optional, we'll fix it then.

DurandA commented:
Hi @shmuelzon.

Thanks for the merge and sorry for the lack of updates. This project is very helpful and I hope more people will benefit from it with WPA-EAP.

I started working on a fork that adds LED status (different blinking pattern depending on if the ESP is connected to the WiFi/broker). It also adds GPS support so that the ESP sends its position to the broker. This is very specific to our project so I don't think it makes sense to merge it. Feel free to take a look anyway (this is not committed yet).
